File: libs/east-py/packages/east-py-io/east_py_io/sql/access.py
#
# Copyright (c) 2025 Elara AI Pty Ltd
# Licensed under the Business Source License 1.1. See LICENSE.md for details.
#
"""Microsoft Access platform functions for East.

Provides read-only Access database operations for East programs, including
opening databases, listing tables, and querying table data with user-defined
return types.

Supported formats:
- .mdb - Access 97, 2000, 2002/2003
- .accdb - Access 2007, 2010, 2013, 2016, 2019

Note: Requires access-parser package. Install with: pip install access-parser
or pip install east-py-io[access]
"""

import importlib.util
import uuid
from datetime import UTC, datetime
from typing import Any

from east.runtime.platform import (
    generic_platform_function,
    platform_function,
    platform_functions,
)
from east.types.types import (
    NullType,
    StringType,
    get_option_inner_type,
    is_blob_type,
    is_boolean_type,
    is_datetime_type,
    is_float_type,
    is_integer_type,
    is_option_type,
    is_string_type,
    is_struct_type,
)
from east.types.values import EastArray, EastStruct, EastVariant

from .types import (
    AccessConfigType,
    AccessTablesResultType,
    ConnectionHandleType,
)

_HAS_ACCESS_SUPPORT = importlib.util.find_spec("access_parser") is not None

# Access type constants (defined locally for use in _ACCESS_TYPE_MAP)
TYPE_BOOLEAN = 1
TYPE_INT8 = 2
TYPE_INT16 = 3
TYPE_INT32 = 4
TYPE_MONEY = 5
TYPE_FLOAT32 = 6
TYPE_FLOAT64 = 7
TYPE_DATETIME = 8
TYPE_BINARY = 9
TYPE_TEXT = 10
TYPE_OLE = 11
TYPE_MEMO = 12
TYPE_GUID = 15


# Connection storage (maps handle -> AccessParser instance)
_access_connections: dict[str, Any] = {}


def _check_access_support() -> None:
    """Check if Access support is available."""
    if not _HAS_ACCESS_SUPPORT:
        raise NotImplementedError(
            "Access support requires the 'access' extra. "
            "Add east-py-io[access] to your pyproject.toml dependencies."
        )


# Access type constant to East type mapping
_ACCESS_TYPE_MAP: dict[int, str] = {
    TYPE_BOOLEAN: "Boolean",
    TYPE_INT8: "Integer",
    TYPE_INT16: "Integer",
    TYPE_INT32: "Integer",
    TYPE_MONEY: "String",  # Money formatted as string
    TYPE_FLOAT32: "Float",
    TYPE_FLOAT64: "Float",
    TYPE_DATETIME: "DateTime",
    TYPE_BINARY: "Blob",
    TYPE_TEXT: "String",
    TYPE_OLE: "Blob",
    TYPE_MEMO: "String",
    TYPE_GUID: "String",
}


def _get_east_type_for_access(access_type: int) -> str:
    """Get East type name for Access column type constant."""
    return _ACCESS_TYPE_MAP.get(access_type, "String")


def _convert_access_value(value: Any, access_type: int) -> Any:
    """Convert an Access value to the appropriate East value."""
    if value is None:
        return None

    if access_type in (TYPE_INT8, TYPE_INT16, TYPE_INT32):
        return int(value)
    elif access_type in (TYPE_FLOAT32, TYPE_FLOAT64):
        return float(value)
    elif access_type == TYPE_BOOLEAN:
        return bool(value)
    elif access_type in (TYPE_BINARY, TYPE_OLE):
        if isinstance(value, (bytes, bytearray)):
            from east.types.values import EastBlob

            return EastBlob(bytes(value))
        return value
    elif access_type == TYPE_DATETIME:
        if isinstance(value, str):
            try:
                value = datetime.fromisoformat(value.replace("Z", "+00:00"))
            except (ValueError, TypeError):
                return None
        if isinstance(value, datetime):
            # Ensure UTC timezone and truncate to milliseconds
            value = value.replace(tzinfo=UTC) if value.tzinfo is None else value.astimezone(UTC)
            ms = (value.microsecond // 1000) * 1000
            return value.replace(microsecond=ms)
        return value
    else:
        # String types - ensure string conversion
        return str(value) if value is not None else value


def _get_east_type_name(t: Any) -> str:
    """Get the East type name from a type object."""
    if is_integer_type(t):
        return "Integer"
    elif is_float_type(t):
        return "Float"
    elif is_boolean_type(t):
        return "Boolean"
    elif is_string_type(t):
        return "String"
    elif is_blob_type(t):
        return "Blob"
    elif is_datetime_type(t):
        return "DateTime"
    return "Unknown"


def _check_type_compatibility(field_type: Any, expected_east: str) -> tuple[bool, bool]:
    """Check if field_type matches expected_east type.

    Returns:
        Tuple of (is_option, is_compatible)
    """
    field_is_option = is_option_type(field_type)
    inner_type = get_option_inner_type(field_type) if field_is_option else field_type
    actual_type = _get_east_type_name(inner_type)

    # Check compatibility
    compatible = False
    if actual_type == expected_east or expected_east == "Integer" and actual_type in ("Float", "Boolean") or expected_east == "Float" and actual_type == "Integer" or expected_east == "Boolean" and actual_type == "Integer":
        compatible = True

    return field_is_option, compatible


@platform_function(
    name="access_open",
    inputs=[AccessConfigType],
    output=ConnectionHandleType,
)
async def access_open_impl(config: EastStruct) -> str:
    """Open a Microsoft Access database file and return a connection handle.

    Supports ``.mdb`` (Access 97/2000/2002/2003) and ``.accdb``
    (Access 2007 and later) formats via access-parser.  Password-protected
    databases are not yet supported by access-parser.

    Args:
        config: ``AccessConfigType`` (``EastStruct``) with fields:

            - ``path`` (``String``): file-system path to the ``.mdb`` or
              ``.accdb`` file.
            - ``password`` (``Option<String>``): database password (reserved -
              not forwarded to access-parser).

    Returns:
        ``String`` - opaque connection handle, passed to
        ``access_tables_impl`` / ``access_query_factory`` /
        ``access_close_impl``.

    Raises:
        NotImplementedError: the ``access`` extra (access-parser) is not
            installed.
        Exception: access-parser fails to open the file (bad path,
            unsupported format, corrupt database).
    """
    _check_access_support()
    from access_parser import AccessParser  # type: ignore[import-untyped]

    try:
        path = config["path"]
        # Note: access-parser doesn't support password-protected databases yet
        # password_opt = config["password"]
        # password = password_opt.value if password_opt.type == "some" else None

        # Open the database
        reader = AccessParser(path)

        handle = str(uuid.uuid4())
        _access_connections[handle] = reader

        return handle
    except Exception as e:
        raise Exception(f"Access database open failed: {e}") from e


@platform_function(
    name="access_tables",
    inputs=[ConnectionHandleType],
    output=AccessTablesResultType,
)
async def access_tables_impl(handle: str) -> EastStruct:
    """List the names of all user tables in the database.

    Args:
        handle: ``String`` - connection handle from ``access_open_impl``.

    Returns:
        ``AccessTablesResultType`` (``EastStruct``): ``{tables:
        Array<String>}`` - table names in catalog order.

    Raises:
        NotImplementedError: the ``access`` extra (access-parser) is not
            installed.
        Exception: the handle is unknown, or access-parser fails to read the
            catalog.
    """
    _check_access_support()

    try:
        if handle not in _access_connections:
            raise Exception(f"Invalid connection handle: {handle}")

        reader = _access_connections[handle]
        # access-parser's catalog is a dict mapping table names to offsets
        tables = list(reader.catalog.keys())

        return EastStruct({"tables": EastArray(StringType, tables)})
    except Exception as e:
        raise Exception(f"Access tables list failed: {e}") from e


def access_query_factory(row_type: Any) -> Any:
    """Return a typed ``access_query`` implementation for a given row struct type.

    Called by the ``@generic_platform_function`` decorator with the resolved
    ``T`` type argument.  The returned coroutine validates Access column types
    against ``T`` using ``_ACCESS_TYPE_MAP`` before converting each row.
    Corrupt memo/overflow records are silently skipped via ``_safe_parse_row``.

    Args:
        row_type: East ``StructType`` describing one result row.

    Returns:
        Async callable ``(handle, options) -> EastArray(T)`` implementing
        ``access_query<T>``.
    """

    async def access_query_impl(handle: str, options: EastStruct) -> EastArray:
        """Query an Access table and return rows typed as ``Array<T>``.

        Validates Access column type codes against ``T`` and applies
        ``rowOffset`` / ``rowLimit`` slicing.  Nullable columns must be
        declared as ``Option<...>`` in ``T``.

        Args:
            handle: ``String`` - connection handle from
                ``access_open_impl``.
            options: ``AccessQueryOptionsType`` (``EastStruct``) with
                fields:

                - ``table`` (``String``): table name.
                - ``columns`` (``Option<Array<String>>``): restrict to these
                  columns (default: all).
                - ``rowOffset`` (``Option<Integer>``): skip this many rows
                  (default 0).
                - ``rowLimit`` (``Option<Integer>``): maximum rows to return
                  (default: all remaining).

        Returns:
            ``Array<T>`` (``EastArray``) - one ``EastStruct`` per result row,
            fields coerced to the types declared in ``T``.

        Raises:
            NotImplementedError: the ``access`` extra (access-parser) is not
                installed.
            Exception: the handle is unknown, the table is not found,
                ``T`` is not a ``StructType``, a required column is missing,
                a column type is incompatible with the corresponding ``T``
                field, or a non-optional field contains ``NULL`` or an
                unparseable value.
        """
        _check_access_support()

        try:
            if handle not in _access_connections:
                raise Exception(f"Invalid connection handle: {handle}")

            reader = _access_connections[handle]

            table_name = options["table"]
            row_offset_opt = options["rowOffset"]
            row_limit_opt = options["rowLimit"]

            row_offset = int(row_offset_opt.value) if row_offset_opt.type == "some" else None
            row_limit = int(row_limit_opt.value) if row_limit_opt.type == "some" else None

            # Get table
            table = reader.get_table(table_name)

            # Validate row type is a Struct
            if not is_struct_type(row_type):
                raise Exception(f"Expected row type must be a Struct, got {row_type.type}")

            fields = row_type.value

            # Get column metadata from table.columns (dict indexed by position)
            # Each column object has: col_name_str (name), type (int constant)
            col_meta: dict[str, Any] = {}
            for col in table.columns.values():
                col_meta[col.col_name_str] = col

            # Validate field types match columns
            field_info: dict[str, dict[str, Any]] = {}
            for field in fields:
                field_name = field["name"]
                field_type = field["type"]

                if field_name not in col_meta:
                    raise Exception(f"Column '{field_name}' not found in table '{table_name}'")

                col = col_meta[field_name]
                access_type = col.type if hasattr(col, "type") else TYPE_TEXT
                expected_east = _get_east_type_for_access(access_type)

                # Check if field type matches expected type or OptionType(expected)
                field_is_option, compatible = _check_type_compatibility(field_type, expected_east)

                if not compatible:
                    raise Exception(
                        f"Type mismatch for column '{field_name}': Access column type is {access_type}, "
                        f"expected {expected_east} or OptionType({expected_east})"
                    )

                field_info[field_name] = {
                    "is_option": field_is_option,
                    "access_type": access_type,
                }

            # Parse table data - returns defaultdict(list) where data[column][row_index]
            # Wrap _parse_row to gracefully handle corrupt memo/overflow records
            # in access-parser (e.g. _parse_memo crashes on None overflow data)
            import contextlib
            original_parse_row = table._parse_row
            def _safe_parse_row(record):
                with contextlib.suppress(Exception):
                    original_parse_row(record)
            table._parse_row = _safe_parse_row
            parsed_data = table.parse()

            # Get the number of rows by checking any column's length
            if not parsed_data:
                return EastArray(row_type, [])

            # Get first column to determine row count
            first_col_data: list = next(iter(parsed_data.values()), [])
            total_rows = len(first_col_data)

            # Apply offset
            start_idx = row_offset if row_offset is not None and row_offset > 0 else 0

            # Apply limit
            end_idx = total_rows
            if row_limit is not None and row_limit >= 0:
                end_idx = min(start_idx + row_limit, total_rows)

            # Convert rows
            rows: list[EastStruct] = []
            for row_idx in range(start_idx, end_idx):
                converted: dict[str, Any] = {}

                for field in fields:
                    field_name = field["name"]
                    info = field_info[field_name]

                    # Get value from parsed data
                    col_data = parsed_data.get(field_name, [])
                    value = col_data[row_idx] if row_idx < len(col_data) else None

                    if value is None:
                        if info["is_option"]:
                            converted[field_name] = EastVariant("none", None)
                        else:
                            raise Exception(
                                f"null value at row[{row_idx}] for required field '{field_name}' - "
                                f"use OptionType for nullable columns"
                            )
                    else:
                        # Convert value based on Access type
                        converted_value = _convert_access_value(value, info["access_type"])

                        if converted_value is None:
                            # Conversion produced None (e.g. unparseable date string)
                            if info["is_option"]:
                                converted[field_name] = EastVariant("none", None)
                            else:
                                raise Exception(
                                    f"invalid value at row[{row_idx}] for required field '{field_name}': {value!r}"
                                )
                        elif info["is_option"]:
                            converted[field_name] = EastVariant("some", converted_value)
                        else:
                            converted[field_name] = converted_value

                rows.append(EastStruct(converted))

            return EastArray(row_type, rows)
        except Exception as e:
            raise Exception(f"Access query failed: {e}") from e

    return access_query_impl


@generic_platform_function(
    name="access_query",
    type_parameters=["T"],
    is_async=True,
)
def _access_query_factory(_platform_list: Any, T: Any) -> Any:  # noqa: N803
    return access_query_factory(T)


@platform_function(
    name="access_close",
    inputs=[ConnectionHandleType],
    output=NullType,
)
async def access_close_impl(handle: str) -> None:
    """Release an Access database connection handle.

    access-parser does not require an explicit close call; this removes the
    internal reference so the parser object can be garbage-collected.

    Args:
        handle: ``String`` - connection handle from ``access_open_impl``.

    Raises:
        Exception: the handle is unknown.
    """
    try:
        if handle not in _access_connections:
            raise Exception(f"Invalid connection handle: {handle}")

        # access-parser doesn't need explicit close, just remove reference
        del _access_connections[handle]
    except Exception as e:
        raise Exception(f"Access close failed: {e}") from e


@platform_function(
    name="access_close_all",
    inputs=[],
    output=NullType,
)
async def access_close_all_impl() -> None:
    """Release every open Access connection handle managed by this process.

    Clears the internal connection map; useful for test teardown.
    """
    _access_connections.clear()


# Collected from the @platform_function / @generic_platform_function decorations above.
access_impl = platform_functions(__name__)

__all__ = [
    "access_impl",
    "access_open_impl",
    "access_tables_impl",
    "access_query_factory",
    "access_close_impl",
    "access_close_all_impl",
]

File: libs/east-node/packages/east-node-io/src/sql/access.ts
/**
 * Copyright (c) 2025 Elara AI Pty Ltd
 * Dual-licensed under AGPL-3.0 and commercial license. See LICENSE for details.
 */

/**
 * Microsoft Access platform functions for East Node IO.
 *
 * Provides read-only Access database operations for East programs, including
 * opening databases, listing tables, and querying table data with user-defined
 * return types.
 *
 * Uses the mdb-reader library which supports Access 97 through Access 2019,
 * including encrypted databases.
 *
 * **Important**: This module is read-only. The mdb-reader library cannot create
 * or modify Access databases.
 *
 * Supported formats:
 * - `.mdb` - Access 97, 2000, 2002/2003
 * - `.accdb` - Access 2007, 2010, 2013, 2016, 2019
 *
 * @packageDocumentation
 */

import { ArrayType, BlobType, BooleanType, DateTimeType, East, FloatType, IntegerType, isTypeValueEqual, NullType, OptionType, StringType, toEastTypeValue, variant } from "@elaraai/east";
import type { EastType, StructTypeValue, ValueTypeOf } from "@elaraai/east";
import type { EastTypeValue, PlatformFunction } from "@elaraai/east/internal";
import { EastError } from "@elaraai/east/internal";
import MDBReader from 'mdb-reader';
import { readFileSync } from 'fs';
import { createHandle, getConnection, closeHandle, closeAllHandles } from '../connection/index.js';
import {
    AccessConfigType,
    AccessQueryOptionsType,
    AccessTablesResultType,
    ConnectionHandleType,
} from './types.js';

/**
 * Opens a Microsoft Access database file.
 *
 * Opens an Access database file (.mdb or .accdb) and returns an opaque handle
 * for use in queries. Supports Access 97 through Access 2019, including
 * encrypted databases.
 *
 * This is a platform function for the East language, enabling Access database
 * read operations in East programs running on Node.js.
 *
 * @param config - Access connection configuration
 * @returns Connection handle (opaque string)
 *
 * @throws {EastError} When connection fails due to:
 * - Invalid file path (location: "access_open")
 * - File not found (location: "access_open")
 * - Invalid password for encrypted database (location: "access_open")
 * - Unsupported database format (location: "access_open")
 *
 * @example
 * ```ts
 * import { East, NullType, variant } from "@elaraai/east";
 * import { SQL } from "@elaraai/east-node-io";
 *
 * const listTables = East.function([], NullType, ($) => {
 *     const config = $.let({
 *         path: "./database.mdb",
 *         password: variant('none', null),
 *     });
 *
 *     const conn = $.let(SQL.Access.open(config));
 *     const tables = $.let(SQL.Access.tables(conn));
 *     $(SQL.Access.close(conn));
 *     $.return(null);
 * });
 *
 * const compiled = East.compileAsync(listTables.toIR(), SQL.Access.Implementation);
 * await compiled();
 * ```
 *
 * @remarks
 * - Access databases are read-only (mdb-reader does not support writes)
 * - Supports encrypted databases with password option
 * - Database file is read entirely into memory on open
 */
export const access_open = East.asyncPlatform("access_open", [AccessConfigType], ConnectionHandleType);

/**
 * Lists all table names in an Access database.
 *
 * Returns an array of table names available in the database.
 * Only returns normal tables (not system or linked tables).
 *
 * @param handle - Connection handle from access_open()
 * @returns Object containing array of table names
 *
 * @throws {EastError} When operation fails due to:
 * - Invalid connection handle (location: "access_tables")
 *
 * @example
 * ```ts
 * import { East, NullType, variant } from "@elaraai/east";
 * import { SQL } from "@elaraai/east-node-io";
 *
 * const showTables = East.function([], NullType, ($) => {
 *     const config = $.let({
 *         path: "./database.mdb",
 *         password: variant('none', null),
 *     });
 *
 *     const conn = $.let(SQL.Access.open(config));
 *     const result = $.let(SQL.Access.tables(conn));
 *     // result.tables contains ['Table1', 'Table2', ...]
 *     $(SQL.Access.close(conn));
 *     $.return(null);
 * });
 *
 * const compiled = East.compileAsync(showTables.toIR(), SQL.Access.Implementation);
 * await compiled();
 * ```
 */
export const access_tables = East.asyncPlatform("access_tables", [ConnectionHandleType], AccessTablesResultType);

/**
 * Queries data from an Access database table with a user-defined return type.
 *
 * Reads rows from a specified table with optional column selection and pagination.
 * The return type is generic, allowing users to specify the expected row structure
 * for type-safe access to query results.
 *
 * This is a generic platform function for the East language, enabling type-safe
 * Access database read operations in East programs running on Node.js.
 *
 * @typeParam T - The expected row type for query results
 * @param handle - Connection handle from access_open()
 * @param options - Query options including table name, columns, offset, and limit
 * @returns Array of rows matching the specified type T
 *
 * @throws {EastError} When query fails due to:
 * - Invalid connection handle (location: "access_query")
 * - Table not found (location: "access_query")
 * - Column not found (location: "access_query")
 *
 * @example
 * ```ts
 * import { East, NullType, StructType, StringType, IntegerType, variant } from "@elaraai/east";
 * import { SQL } from "@elaraai/east-node-io";
 *
 * // Define the expected row type
 * const UserRowType = StructType({
 *     id: IntegerType,
 *     name: StringType,
 *     email: StringType,
 * });
 *
 * const queryUsers = East.function([], NullType, ($) => {
 *     const config = $.let({
 *         path: "./database.mdb",
 *         password: variant('none', null),
 *     });
 *
 *     const conn = $.let(SQL.Access.open(config));
 *     // Query with typed results - returns Array<UserRowType>
 *     const users = $.let(SQL.Access.query([UserRowType], conn, {
 *         table: "Users",
 *         columns: variant('some', ["id", "name", "email"]),
 *         rowOffset: variant('none', null),
 *         rowLimit: variant('some', 100n),
 *     }));
 *     // users is typed as Array<{ id: bigint, name: string, email: string }>
 *     $(SQL.Access.close(conn));
 *     $.return(null);
 * });
 *
 * const compiled = East.compileAsync(queryUsers.toIR(), SQL.Access.Implementation);
 * await compiled();
 * ```
 *
 * @remarks
 * - Table names are case-sensitive
 * - The row type T should match the structure of the queried columns
 * - All Access data types are mapped to appropriate East types
 * - NULL values are preserved in the result
 */
export const access_query = East.asyncGenericPlatform(
    "access_query",
    ["T"],
    [ConnectionHandleType, AccessQueryOptionsType],
    ArrayType("T")
);

/**
 * Closes an Access database connection.
 *
 * Closes the database connection and releases all resources.
 * The handle becomes invalid after this operation.
 *
 * @param handle - Connection handle from access_open()
 *
 * @throws {EastError} When handle is invalid (location: "access_close")
 *
 * @example
 * ```ts
 * import { East, NullType, variant } from "@elaraai/east";
 * import { SQL } from "@elaraai/east-node-io";
 *
 * const cleanup = East.function([], NullType, $ => {
 *     const config = $.let({
 *         path: "./database.mdb",
 *         password: variant('none', null),
 *     });
 *     const conn = $.let(SQL.Access.open(config));
 *     // ... do work ...
 *     $(SQL.Access.close(conn));
 *     return $.return(null);
 * });
 *
 * const compiled = East.compileAsync(cleanup.toIR(), SQL.Access.Implementation);
 * await compiled();
 * ```
 */
export const access_close = East.asyncPlatform("access_close", [ConnectionHandleType], NullType);

/**
 * Closes all Access database connections.
 *
 * Closes all active Access connections and releases all resources.
 * Useful for test cleanup to ensure all connections are closed.
 *
 * @returns Null on success
 *
 * @example
 * ```ts
 * import { East, NullType } from "@elaraai/east";
 * import { SQL } from "@elaraai/east-node-io";
 *
 * const cleanupAll = East.function([], NullType, $ => {
 *     // ... test code that may have left connections open ...
 *     $(SQL.Access.closeAll());
 *     return $.return(null);
 * });
 *
 * const compiled = East.compileAsync(cleanupAll.toIR(), SQL.Access.Implementation);
 * await compiled();
 * ```
 *
 * @internal
 */
export const access_close_all = East.asyncPlatform("access_close_all", [], NullType);

/**
 * Node.js implementation of Access platform functions.
 *
 * Provides the runtime implementations for Access database operations.
 * Pass this to East.compileAsync() to enable Access functionality.
 */
export const AccessImpl: PlatformFunction[] = [
    access_open.implement(async (config: ValueTypeOf<typeof AccessConfigType>): Promise<string> => {
        try {
            const path = config.path;
            const password = config.password?.type === 'some' ? config.password.value : undefined;

            // Read the database file into a buffer
            const buffer = readFileSync(path);

            // Create the MDB reader with optional password
            const readerOptions = password ? { password } : undefined;
            const reader = new MDBReader(buffer, readerOptions);

            return await Promise.resolve(createHandle(reader, async () => {
                // mdb-reader doesn't have a close method, but we clean up the handle
                await Promise.resolve();
            }));
        } catch (err: any) {
            throw new EastError(`Access database open failed: ${err.message}`, {
                location: [{ filename: "access_open", line: 0n, column: 0n }],
                cause: err
            });
        }
    }),

    access_tables.implement(async (handle: ValueTypeOf<typeof ConnectionHandleType>): Promise<ValueTypeOf<typeof AccessTablesResultType>> => {
        try {
            const reader = await Promise.resolve(getConnection<MDBReader>(handle));

            // Get only normal tables (not system or linked tables)
            const tables = reader.getTableNames({
                normalTables: true,
                systemTables: false,
                linkedTables: false
            });

            return { tables };
        } catch (err: any) {
            throw new EastError(`Access tables list failed: ${err.message}`, {
                location: [{ filename: "access_tables", line: 0n, column: 0n }],
                cause: err
            });
        }
    }),

    access_query.implement((T: EastTypeValue) => async (...args: unknown[]): Promise<unknown> => {
        const handle = args[0] as string;
        const options = args[1] as ValueTypeOf<typeof AccessQueryOptionsType>;
        const rowType = T as EastType;

        try {
            const reader = await Promise.resolve(getConnection<MDBReader>(handle));

            const tableName = options.table;
            const columns = options.columns?.type === 'some' ? options.columns.value : undefined;
            const rowOffset = options.rowOffset?.type === 'some' ? Number(options.rowOffset.value) : undefined;
            const rowLimit = options.rowLimit?.type === 'some' ? Number(options.rowLimit.value) : undefined;

            // Get the table
            const table = reader.getTable(tableName);

            // Get column metadata for type validation and conversion
            const columnDefs = table.getColumns();
            const columnMetaMap = new Map<string, { type: string; nullable: boolean }>();
            for (const col of columnDefs) {
                columnMetaMap.set(col.name, { type: col.type, nullable: col.nullable });
            }

            // Validate that the expected row type matches the Access column metadata
            if (rowType.type !== 'Struct') {
                throw new EastError(
                    `Expected row type must be a Struct, got ${rowType.type}`,
                    { location: [{ filename: "access_query", line: 0n, column: 0n }] }
                );
            }

            // Note: T is EastTypeValue (IR format), not EastType (runtime format)
            const structTypeValue = rowType as unknown as StructTypeValue;
            for (const field of structTypeValue.value) {
                const fieldName = field.name;
                const fieldType = field.type;
                const colMeta = columnMetaMap.get(fieldName);
                if (!colMeta) {
                    throw new EastError(
                        `Column '${fieldName}' not found in table '${tableName}'`,
                        { location: [{ filename: "access_query", line: 0n, column: 0n }] }
                    );
                }

                // Determine expected East type based on Access column type
                let expectedBaseType: EastType;
                switch (colMeta.type) {
                    case 'boolean':
                        expectedBaseType = BooleanType;
                        break;
                    case 'byte':
                    case 'integer':
                    case 'long':
                    case 'bigint':
                        expectedBaseType = IntegerType;
                        break;
                    case 'float':
                    case 'double':
                        expectedBaseType = FloatType;
                        break;
                    case 'text':
                    case 'memo':
                    case 'currency':
                    case 'numeric':
                    case 'repid':
                    case 'datetimeextended':
                        expectedBaseType = StringType;
                        break;
                    case 'datetime':
                        expectedBaseType = DateTimeType;
                        break;
                    case 'binary':
                    case 'ole':
                        expectedBaseType = BlobType;
                        break;
                    default:
                        throw new EastError(
                            `Unsupported Access column type '${colMeta.type}' for column '${fieldName}'`,
                            { location: [{ filename: "access_query", line: 0n, column: 0n }] }
                        );
                }

                // Check type matches, accounting for nullable columns
                const expectedType = colMeta.nullable ? OptionType(expectedBaseType) : expectedBaseType;
                if (!isTypeValueEqual(fieldType, toEastTypeValue(expectedType))) {
                    const expectedTypeName = colMeta.nullable
                        ? `OptionType(${expectedBaseType.type})`
                        : expectedBaseType.type;
                    throw new EastError(
                        `Type mismatch for column '${fieldName}': Access column is ${colMeta.type}${colMeta.nullable ? ' (nullable)' : ''}, got ${fieldType.type} but expected East type ${expectedTypeName}`,
                        { location: [{ filename: "access_query", line: 0n, column: 0n }] }
                    );
                }
            }

            // Get data with options
            const dataOptions: { columns?: string[]; rowOffset?: number; rowLimit?: number } = {};
            if (columns) dataOptions.columns = columns;
            if (rowOffset !== undefined) dataOptions.rowOffset = rowOffset;
            if (rowLimit !== undefined) dataOptions.rowLimit = rowLimit;

            const rawData = table.getData(dataOptions);

            // Convert mdb-reader types to East types based on column metadata
            // mdb-reader returns: integer/long/byte → number, binary/ole → Buffer
            // East expects: IntegerType → bigint, BlobType → Uint8Array
            const data = rawData.map((row: Record<string, any>) => {
                const converted: Record<string, any> = {};
                for (const [key, value] of Object.entries(row)) {
                    const colMeta = columnMetaMap.get(key);
                    if (value === null) {
                        // For nullable columns, wrap null in the 'none' variant
                        converted[key] = colMeta?.nullable ? variant('none', null) : null;
                    } else {
                        let convertedValue: any;
                        switch (colMeta?.type) {
                            case 'integer':
                            case 'long':
                            case 'byte':
                                // Convert number to bigint for East IntegerType
                                convertedValue = BigInt(value as number);
                                break;
                            case 'binary':
                            case 'ole':
                                // Convert Buffer to Uint8Array for East BlobType
                                convertedValue = new Uint8Array(value as Buffer);
                                break;
                            default:
                                // datetime → Date, boolean → boolean, float/double → number,
                                // text/memo → string, bigint → bigint (already correct)
                                convertedValue = value;
                        }
                        // For nullable columns, wrap value in the 'some' variant
                        converted[key] = colMeta?.nullable ? variant('some', convertedValue) : convertedValue;
                    }
                }
                return converted;
            });

            // Type validation is already done field-by-field above during column metadata validation
            return data;
        } catch (err: any) {
            if (err instanceof EastError) {
                throw err;
            }
            throw new EastError(`Access query failed: ${err.message}`, {
                location: [{ filename: "access_query", line: 0n, column: 0n }],
                cause: err
            });
        }
    }),

    access_close.implement(async (handle: ValueTypeOf<typeof ConnectionHandleType>) => {
        try {
            // Verify handle exists (mdb-reader doesn't need explicit close)
            await Promise.resolve(getConnection<MDBReader>(handle));
            closeHandle(handle);
            return null;
        } catch (err: any) {
            throw new EastError(`Access close failed: ${err.message}`, {
                location: [{ filename: "access_close", line: 0n, column: 0n }],
                cause: err
            });
        }
    }),

    access_close_all.implement(async () => {
        await closeAllHandles();
        return null;
    }),
];
